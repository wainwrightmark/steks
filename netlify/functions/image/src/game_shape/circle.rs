use std::format;

use super::GameShapeBody;

#[derive(PartialEq, Eq, Clone, Copy, Debug, Hash, Default)]
pub struct Circle;

impl GameShapeBody for Circle {
    fn as_svg(&self, size: f32, color_rgba: String) -> String {
        let size = size * std::f32::consts::FRAC_2_SQRT_PI * 0.5;
        format!(r#"<circle r="{size}" fill="{color_rgba}" stroke-width="0" />"#)
    }
}
