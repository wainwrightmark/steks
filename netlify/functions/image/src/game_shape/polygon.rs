use std::format;

use super::GameShapeBody;

#[derive(PartialEq, Eq, Clone, Copy, Debug, Hash)]
pub struct PolygonBody<const SQUARES: usize, const POINTS: usize>(
    pub &'static [(isize, isize); POINTS],
);

impl<const S: usize, const P: usize> GameShapeBody for PolygonBody<S, P> {
    fn as_svg(&self, size: f32, color_rgba: String) -> String {
        let u = size / (1.0 * f32::sqrt(S as f32));

        let points = self
            .0
            .map(|(x, y)| format!("{},{}", (x as f32) * u, (y as f32) * u))
            .join(" ");

        format!(r#"<polygon points="{points}" fill="{color_rgba}" />"#)
    }
}
