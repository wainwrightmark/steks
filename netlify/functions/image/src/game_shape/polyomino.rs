// use super::{GameShapeBody, SHAPE_RADIUS};
// use crate::{
//     grid::prelude::{PolyominoShape, Shape},
//     PHYSICS_SCALE,
// };
use geometrid::{
    polyomino::Polyomino,
    prelude::{HasCenter, Location},
    shape::Shape,
};

use crate::Vec2;

use super::GameShapeBody;

fn get_vertices<const S: usize>(
    shape: &Polyomino<S>,
    shape_size: f32,
) -> impl Iterator<Item = Vec2> {
    let u = shape_size / (1.0 * f32::sqrt(S as f32));
    let Location {
        x: x_offset,
        y: y_offset,
    } = shape.get_center(1.0);

    shape.draw_outline().map(move |qr| {
        Vec2::new(
            ((qr.x as f32) - x_offset) * u,
            ((qr.y as f32) - y_offset) * u,
        )
    })
}
impl<const S: usize> GameShapeBody for Polyomino<S> {
    fn as_svg(&self, size: f32, color_rgba: String) -> String {
        let points: Vec<_> = get_vertices(&self, size)
            .map(|v| format!("{},{}", v.x, v.y))
            .collect();

        let points = points.join(" ");

        format!(r#"<polygon points="{points}" fill="{color_rgba}" />"#)
    }
}
