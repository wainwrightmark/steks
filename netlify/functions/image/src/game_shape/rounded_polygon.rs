use geometrid::prelude::Point;

/// Adds a sub-path from a polygon but rounds the corners.
///
/// There must be no sub-path in progress when this method is called.
/// No sub-path is in progress after the method is called.
pub fn make_rounded_polygon_path(points: &[Point], radius: f32)-> String {
    if points.len() < 2 {
        return "".to_string();
    }

    //p points are original polygon points
    //q points are the actual points we will draw lines and arcs between
    let clamped_radius = clamp_radius(radius, points[points.len() - 1], points[0], points[1]);
    let q_first = get_point_between(points[0], points[1], clamped_radius);

    //We begin on the line just after the first point
    let mut builder = format!("M{} {}", q_first.x, q_first.y);
    //builder.begin(q_first, attributes);

    for index in 0..points.len() {
        let p_current = points[index];
        let p_next = points[(index + 1) % points.len()];
        let p_after_next = points[(index + 2) % points.len()];

        let clamped_radius = clamp_radius(radius, p_current, p_next, p_after_next);

        //q1 is the second point on the line between p_current and p_next
        let q1 = get_point_between(p_next, p_current, clamped_radius);
        //q2 is the first point on the line between p_next and p_after_next
        let q2 = get_point_between(p_next, p_after_next, clamped_radius);

        line_to(&mut builder, q1);
        let turn_winding = get_winding(p_current, p_next, p_after_next);

        //Draw the arc near p_next
        arc(
            &mut builder,
            (clamped_radius, clamped_radius),
            0.0,
            ArcFlags {
                large_arc: false,
                sweep: turn_winding,
            },
            q2,
        );
    }

    builder.push('z');

    builder
}

#[derive(Debug, Copy, Clone)]
struct ArcFlags {
    pub large_arc: bool,
    pub sweep: Winding,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq)]
enum Winding {
    Positive,
    Negative,
}

fn clamp_radius(radius: f32, p_previous: Point, p_current: Point, p_next: Point) -> f32 {
    let shorter_edge = ((p_current - p_next).length()).min((p_previous - p_current).length());
    let clamped_radius = radius.min(shorter_edge * 0.5);
    clamped_radius
}

fn get_point_between(p1: Point, p2: Point, radius: f32) -> Point {
    let dist = p1.distance_to(&p2);
    let ratio = radius / dist;

    p1.lerp(p2, ratio)
}

fn get_winding(p0: Point, p1: Point, p2: Point) -> Winding {
    let cross = (p2 - p0).cross_product(p1 - p0);
    if cross.is_sign_positive() {
        Winding::Positive
    } else {
        Winding::Negative
    }
}

fn line_to(builder: &mut String, to: Point) {
    builder.push_str(format!("L {} {}", to.x, to.y).as_str())
}

fn arc( builder:&mut String, radii: (f32, f32), x_rotation: f32, flags: ArcFlags, to: Point) {
    let rx = radii.0;
    let ry = radii.1;
    let large_arc = flags.large_arc.then(|| 1).unwrap_or_default();
    let sweep = if flags.sweep == Winding::Negative {
        1
    } else {
        0
    };
    let x = to.x;
    let y = to.y;

    builder.push_str(format!("A {rx} {ry} {x_rotation} {large_arc} {sweep} {x} {y}").as_str())
}
