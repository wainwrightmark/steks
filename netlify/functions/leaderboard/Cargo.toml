[package]
name = "leaderboard"
version = "0.1.0"
edition = "2021"
rust-version = "1.68.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
aws_lambda_events = {version= "0.9", default-features=false, features=["apigw"]}
lambda_runtime = {version= "*", default-features=false}
tokio = {version= "*", default-features=false}
itertools = "0.10.5"
planetscale-driver =  {version="0.4", default-features=false }
reqwest= {version= "0.11.17", default-features=false, features=["rustls-tls"]}
anyhow = "1.0.71"
