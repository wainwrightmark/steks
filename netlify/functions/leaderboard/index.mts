import type { Context } from "@netlify/functions"
import { getStore } from "@netlify/blobs";


function set_headers(response : Response){
    response.headers.set('Access-Control-Allow-Origin', '*');
    response.headers.set('Access-Control-Allow-Methods', "GET, POST");
    response.headers.set('Access-Control-Allow-Headers', '*');
}

export default async (req: Request, context: Context) => {

    const url = new URL(req.url);
    const command = url.searchParams.get("command");
    const store = getStore({ name: "leaderboard", consistency: "strong" });


    switch (command?.toLowerCase()){
        case "get": {
            //we apparently don't ever use this

            const response = new Response("");

            set_headers(response);
            return response;

        }

        case "getrow":{
            const hash : string | null  = url.searchParams.get("hash");

            if (hash ){
                const data : string | null = await store.get(hash);

                if (data === null){
                    return new Response(`${hash} 0.0 0`);
                }

                const response = new Response(`${hash} ${data}`);

                set_headers(response);

                return response;
            }
            else{
                throw new Error("Could not get hash");
            }
        }

        case "set":{
            const hash : string | null  = url.searchParams.get("hash");
            const height : string  = url.searchParams.get("height")??"0.0";
            const height_f: number = parseFloat(height);
            const blob : string  = url.searchParams.get("blob")??"0";

            if (hash ){
                const data : string | null = await store.get(hash);

                if (data === null){

                    await store.set(hash, `${height} ${blob}`);
                    return new Response();
                }

                const split = data.split(" ");


                const prev_height = split[0];


                const prev_height_f : number = parseFloat(prev_height);

                if (height_f > prev_height_f){

                    await store.set(hash, `${height} ${blob}`);
                }

                const response = new Response();

                set_headers(response);

                return response;
            }
            else{
                throw new Error("Could not get hash");
            }
        }
        default:{
            throw new Error("Could not get command");
        }
    }
}
