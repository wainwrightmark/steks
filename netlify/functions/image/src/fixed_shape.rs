use crate::game_shape::*;
use crate::*;
use crate::point::Point;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct FixedShape {
    pub shape: &'static GameShape,
    pub fixed_location: Location,
    pub locked: bool,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Location {
    pub position: Point,
    pub angle: f32,
}

impl Location {
    pub fn svg_transform(&self) -> String {
        let x = self.position.x;
        let y = self.position.y;
        let deg = self.angle.to_degrees();
        format!("translate({x} {y}) rotate({deg})  ")
    }
}

impl FixedShape {
    pub fn with_location(mut self, position: Point, angle: f32) -> Self {
        self.fixed_location = Location { position, angle };
        self
    }

    pub fn lock(mut self) -> Self {
        self.locked = true;
        self
    }
}
