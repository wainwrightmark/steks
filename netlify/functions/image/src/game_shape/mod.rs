use once_cell::sync::Lazy;
use std::fmt::Debug;

use crate::color::choose_color;

use geometrid::polyomino::Polyomino;

pub mod circle;

pub mod polygon;
pub mod polyomino;
pub mod rounded_polygon;

pub use circle::*;

pub use polygon::*;
use resvg::usvg::Color;

pub trait GameShapeBody: Send + Sync {
    fn as_svg(&self, size: f32, color_rgba: String) -> String;
}

const SHAPE_RADIUS_RATIO: f32 = 0.1;
#[derive(Clone)]
pub struct GameShape {
    pub name: &'static str,
    pub body: &'static dyn GameShapeBody,
    pub index: usize,
}

impl Eq for GameShape {}

impl PartialEq for GameShape {
    fn eq(&self, other: &Self) -> bool {
        self.index == other.index
    }
}

impl GameShape {
    pub fn default_fill_color(&self) -> Color {
        choose_color(self.index)
    }

    pub fn fill(&self) -> Color {
        self.default_fill_color()
    }
}

impl Debug for GameShape {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl std::fmt::Display for GameShape {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

pub static ALL_SHAPES: Lazy<Vec<GameShape>> = Lazy::new(|| {
    let v1: [(&'static dyn GameShapeBody, &'static str); 2] =
        [(&Circle {}, "Circle"), (&TRIANGLE, "Triangle")];

    let tetrominos = Polyomino::TETROMINOS
        .iter()
        .map(|x| x as &'static dyn GameShapeBody)
        .zip(Polyomino::TETROMINO_NAMES);
    let pentominos = Polyomino::FREE_PENTOMINOS
        .iter()
        .map(|x| x as &'static dyn GameShapeBody)
        .zip(Polyomino::FREE_PENTOMINO_NAMES);

    v1.into_iter()
        .chain(tetrominos)
        .chain(pentominos)
        .enumerate()
        .map(|(index, (body, name))| GameShape { name, body, index })
        .collect()
});

pub fn shape_by_name(name: &'static str) -> Option<&GameShape> {
    ALL_SHAPES.iter().find(|x| x.name == name)
}

const TRIANGLE: PolygonBody<4, 3> = PolygonBody(&[(-1, -1), (-1, 2), (2, -1)]);
