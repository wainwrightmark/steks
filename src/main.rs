use std::time::Duration;

use app_redirect::AppUrlPlugin;
use bevy::log::*;
use bevy::prelude::*;

use bevy::window::WindowResizeConstraints;
use bevy::window::WindowResolution;
use bevy_embedded_assets::EmbeddedAssetPlugin;

use bevy_pkv::PkvStore;
use bevy_prototype_lyon::prelude::*;
use bevy_rapier2d::prelude::*;

pub const WINDOW_WIDTH: f32 = 360f32;

pub const WINDOW_HEIGHT: f32 = 520f32;

//Be aware that changing these will mess with the saved and shared data
pub const MAX_WINDOW_WIDTH: f32 = 1920f32;
pub const MAX_WINDOW_HEIGHT: f32 = 1080f32;

pub const WALL_WIDTH: f32 = 1920f32;

pub const PHYSICS_SCALE: f32 = 64f32;
mod camera;
mod color;
mod draggable;
pub mod encoding;
pub mod fixed_shape;
pub mod padlock;
mod saved_data;
pub mod set_level;
pub mod share;

pub mod app_redirect;

pub mod level_ui;

pub mod shapes_vec;

pub mod infinity;

pub mod async_event_writer;

use fireworks::FireworksPlugin;
use lens::LensPlugin;
use level_ui::LevelUiPlugin;
//use menu_action::MenuActionPlugin;
use padlock::*;

use bevy_tweening::TweeningPlugin;
use camera::*;
use draggable::*;
//use recording::RecordingPlugin;
use saved_data::*;
mod level;
use level::*;
mod walls;
use share::SharePlugin;
use spirit::SpiritPlugin;
use walls::*;

mod shape_maker;

mod menu;
use menu::*;

mod win;
use win::*;

mod input;
use input::*;

mod collision;
use collision::*;

mod leaderboard;
use leaderboard::*;

mod fireworks;

mod spirit;

mod notifications;

mod game_shape;
use fixed_shape::*;
use game_shape::*;
use notifications::*;

#[cfg(target_arch = "wasm32")]
use crate::logging::LoggableEvent;

#[cfg(target_arch = "wasm32")]
mod logging;

//pub const ZOOM_ENTITY_LAYER: u8 = 1;

#[cfg(target_arch = "wasm32")]
mod wasm;

mod recording;

mod lens;

fn main() {
    // When building for WASM, print panics to the browser console
    #[cfg(target_arch = "wasm32")]
    console_error_panic_hook::set_once();

    let window_plugin = WindowPlugin {
        primary_window: Some(Window {
            title: "steks".to_string(),
            canvas: Some("#game".to_string()),
            resolution: WindowResolution::new(WINDOW_WIDTH, WINDOW_HEIGHT),
            resize_constraints: WindowResizeConstraints {
                min_height: WINDOW_HEIGHT,
                min_width: WINDOW_WIDTH,
                max_width: MAX_WINDOW_WIDTH,
                max_height: MAX_WINDOW_HEIGHT,
            },
            present_mode: bevy::window::PresentMode::default(),

            resizable: true,
            ..Default::default()
        }),
        ..Default::default()
    };

    let log_plugin = LogPlugin {
        level: Level::INFO,
        ..Default::default()
    };
    let mut builder = App::new();

    builder
        .insert_resource(Msaa::Sample4)
        .insert_resource(ClearColor(color::BACKGROUND_COLOR))
        .add_plugins(
            DefaultPlugins
                .set(window_plugin)
                .set(log_plugin)
                .build()
                .add_before::<bevy::asset::AssetPlugin, _>(EmbeddedAssetPlugin),
        )
        .add_plugin(WallsPlugin)
        .add_plugin(ButtonPlugin)
        .add_plugin(ShapePlugin)
        .add_plugin(InputPlugin)
        .add_plugin(CameraPlugin)
        .add_plugin(LeaderboardPlugin)
        .add_plugin(SpiritPlugin)
        .add_plugin(LevelUiPlugin)
        .add_plugin(LensPlugin)
        .add_plugin(FireworksPlugin)
        .add_plugin(NotificationPlugin)
        .add_plugin(AppUrlPlugin)
        //.add_plugin(MenuActionPlugin)
        .add_plugin(RapierPhysicsPlugin::<NoUserData>::pixels_per_meter(
            PHYSICS_SCALE,
        ))
        .add_startup_system(setup)
        .add_plugin(DragPlugin)
        .add_plugin(WinPlugin)
        .add_plugin(LevelPlugin)
        .add_plugin(TweeningPlugin)
        .add_plugin(SharePlugin)
        .add_plugin(CollisionPlugin)
        .add_plugin(PadlockPlugin)
        //.add_plugin(RecordingPlugin)
        .insert_resource(PkvStore::new("Wainwrong", "steks"))
        .insert_resource(bevy::winit::WinitSettings {
            return_from_run: false,
            focused_mode: bevy::winit::UpdateMode::Continuous,
            unfocused_mode: bevy::winit::UpdateMode::ReactiveLowPower {
                max_wait: Duration::from_secs(60),
            },
        });

    #[cfg(target_arch = "wasm32")]
    builder.add_plugin(wasm::WASMPlugin);

    if cfg!(debug_assertions) {
        //builder.add_plugin(RapierDebugRenderPlugin::default());
        //builder.add_plugin(ScreenDiagsPlugin);
        // builder.add_plugin(bevy::diagnostic::LogDiagnosticsPlugin::default());
        // builder.add_plugin(bevy::diagnostic::FrameTimeDiagnosticsPlugin::default());
    }

    builder.add_startup_system(disable_back);
    builder.add_startup_system(hide_splash);
    builder.add_startup_system(set_status_bar.after(hide_splash));
    builder.add_startup_system(log_start.in_base_set(StartupSet::PostStartup));
    builder.run();
}

pub fn setup(mut rapier_config: ResMut<RapierConfiguration>) {
    rapier_config.gravity = GRAVITY;
}

// About 400 is a good amount of wind
pub const GRAVITY: Vec2 = Vec2::new(0.0, -1000.0);

pub fn get_today_date() -> chrono::NaiveDate {
    let today = chrono::offset::Utc::now();
    today.date_naive()
}

pub fn log_start(mut pkv: ResMut<PkvStore>) {
    const KEY: &str = "UserExists";

    let user_exists = pkv.get::<bool>(KEY).ok().unwrap_or_default();

    if !user_exists {
        pkv.set(KEY, &true).unwrap();
    }

    bevy::tasks::IoTaskPool::get()
        .spawn(async move { log_start_async(user_exists).await })
        .detach();
}

fn disable_back() {
    bevy::tasks::IoTaskPool::get()
        .spawn(async move { disable_back_async().await })
        .detach();
}

fn hide_splash() {
    #[cfg(all(target_arch = "wasm32", any(feature = "android", feature = "ios")))]
    {
        bevy::tasks::IoTaskPool::get()
            .spawn(
                async move { capacitor_bindings::splash_screen::SplashScreen::hide(5000.0).await },
            )
            .detach();
    }
}

fn set_status_bar() {
    #[cfg(all(target_arch = "wasm32", any(feature = "android", feature = "ios")))]
    {
        use capacitor_bindings::status_bar::*;
        bevy::tasks::IoTaskPool::get()
            .spawn(async move {
                logging::do_or_report_error_async(|| StatusBar::set_style(Style::Dark)).await;
                logging::do_or_report_error_async(|| StatusBar::set_background_color("#5B8BE2"))
                    .await;
                //logging::do_or_report_error_async(|| StatusBar::hide()).await;
            })
            .detach();
    }
}

async fn disable_back_async<'a>() {
    #[cfg(all(feature = "android", target_arch = "wasm32"))]
    {
        let result = capacitor_bindings::app::App::add_back_button_listener(|_| {
            bevy::tasks::IoTaskPool::get()
                .spawn(async move {
                    logging::do_or_report_error_async(|| {
                        capacitor_bindings::app::App::minimize_app()
                    })
                    .await;
                })
                .detach();
        })
        .await;

        match result {
            Ok(handle) => {
                handle.leak();
            }
            Err(err) => {
                crate::logging::try_log_error_message(format!("{err}"));
            }
        }
    }
}

async fn log_start_async<'a>(_user_exists: bool) {
    #[cfg(target_arch = "wasm32")]
    {
        let device_id = match capacitor_bindings::device::Device::get_id().await {
            Ok(device_id) => device_id,
            Err(err) => {
                crate::logging::try_log_error_message(format!("{err:?}"));
                return;
            }
        };

        if !_user_exists {
            let new_user = wasm::new_user_async().await;
            new_user.try_log_async1(device_id.clone()).await;
        }
        let application_start = wasm::application_start().await;
        application_start.try_log_async1(device_id).await;
    }
}
