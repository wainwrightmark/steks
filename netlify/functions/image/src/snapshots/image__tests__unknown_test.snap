---
source: src/main.rs
expression: hash
---
17216766335305622750
